cdap-app-fabric/src/main/java/co/cask/cdap/app/DefaultAppConfigurer.java

```diff
@@ -277,14 +276,18 @@ public ApplicationSpecification createSpecification(@Nullable String application
       ? ArtifactScope.SYSTEM : ArtifactScope.USER;
     ArtifactId artifactId = new ArtifactId(this.artifactId.getName(), this.artifactId.getVersion(), scope);
 
+    String namespace = deployNamespace.toEntityId().getNamespace();
     String appName = applicationName == null ? name : applicationName;
     String appVersion = applicationVersion == null ? ApplicationId.DEFAULT_VERSION : applicationVersion;
 
-    Map<String, ScheduleCreationSpec> scheduleSpecs = new HashMap<>();
-    for (Map.Entry<String, ScheduleCreationBuilder> entry : programSchedules.entrySet()) {
-      String scheduleName = entry.getKey();
-      ScheduleCreationBuilder builder = entry.getValue();
-      scheduleSpecs.put(scheduleName, builder.build(deployNamespace.toEntityId().getNamespace(), appName, appVersion));
+    for (Map.Entry<String, ScheduleCreationSpec> entry : scheduleSpecs.entrySet()) {
+      // If the ScheduleCreationSpec is really a builder, then build the ScheduleCreationSpec
+      if (entry.getValue() instanceof ScheduleBuilder.ScheduleCreationBuilder) {
+        String scheduleName = entry.getKey();
+        ScheduleBuilder.ScheduleCreationBuilder builder = (ScheduleBuilder.ScheduleCreationBuilder) entry.getValue();
+        // Overwrite the ScheduleCreationBuilder with a built, immutable ScheduleCreationSpec
+        scheduleSpecs.put(scheduleName, builder.build(namespace, appName, appVersion));
```

Comment:
I would rather not to modify the same map while iterating. It's better to just copying everything to a new map.